### PR TITLE
Feature: various things to improve output

### PR DESCRIPTION
**Pull request by liss-h**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -16,48 +17,80 @@ use update_worker::UpdateWorker;
...
         #[clap(short = 'w', long)]
         num_update_workers: usize,
 
+        /// Path to the directory that contains the information for the updaters
         #[clap(short = 'Q', long)]
         update_query_dir: PathBuf,
 
-        #[clap(short = 't', long, default_value_t = 0)]
-        runtime_s: u64,
-
```
Comment (nkaralis) [inline src/main.rs:52]: was this removed because a time limit does not make sense in verification workloads?

File: src/main.rs [hunk 2]
```diff
@@ -203,18 +233,18 @@ async fn main() -> anyhow::Result<()> {
             WorkerType::Reader => {
                 n_readers_finished += 1;
 
-                if let Ok((query_timings, total_time)) = res {
-                    let qps = query_timings.len() as f64 / total_time.as_secs_f64();
+                if let Ok(query_timings) = res {
+                    let reader_qps = query_timings.values().sum::<f64>() / query_timings.len() as f64;
```
Comment (nkaralis) [inline src/main.rs]: do you actually store the qps or avgqps here?

Comment (nkaralis) [inline src/main.rs]: what does query_timings.values() store?

File: src/main.rs [hunk 3]
```diff
@@ -225,12 +255,10 @@ async fn main() -> anyhow::Result<()> {
     }
 
     tracing::info!(
-        "The readers achieved {} AvgQPS / {} IGUANA AvgQPS",
+        "The readers achieved {} AvgQPS",
```
Comment (nkaralis) [inline src/main.rs]: This is the average of AvgQPS across all readers, right?

File: src/random_read_worker.rs [hunk 4]
```diff
@@ -94,7 +98,9 @@ impl RandomReadWorker {
 
         success?;
 
-        let end_time = Instant::now();
-        Ok((query_times, end_time.duration_since(start_time)))
+        Ok(query_qps
```
Comment (nkaralis) [inline src/random_read_worker.rs]: what's the formula for qps? adding it as a comment would be helpful

File: src/random_read_worker.rs [hunk 5]
```diff
@@ -31,26 +29,29 @@ impl FileSourceQueryGenerator {
...
             self.queries.shuffle(&mut rand::thread_rng());
         }
 
         self.ix = (self.ix + 1) % self.queries.len();
 
-        Cow::Borrowed(&self.queries[cur_ix])
+        let ret_query = &self.queries[cur_ix];
+        let orig_ix = self.queries_original_order.iter().position(|q| q == ret_query).unwrap();
+
+        (Some(orig_ix), Cow::Borrowed(ret_query))
```
Comment (nkaralis) [inline src/random_read_worker.rs:62]: do all workers now follow the same order?

**Review by nkaralis — COMMENTED** [review]
I left some comments and questions. They might make the code a bit easier to follow at those points.

File: src/main.rs (hunk 1 above)
Comment (liss-h) [reply src/main.rs:52]: Yes, and it was unused anyway

File: src/main.rs (hunk 2 above)
Comment (liss-h) [reply src/main.rs]: just the values of the (query id -> avgqps) map; So this is a list of the avgqps for all queries

Comment (liss-h) [reply src/main.rs]: avgqps, will rename vars

Comment (nkaralis) [reply src/main.rs]: Do you mean the QPS of all queries?

Comment (liss-h) [reply src/main.rs]: for each query its the average of all QPS measurements for that query (as each query might be run multiple times)

File: src/random_read_worker.rs (hunk 5 above)
Comment (liss-h) [reply src/random_read_worker.rs:62]: They all associate the same IDs with the queries, but they do not execute them in the same order.

Basically:
1. read file, store queries in original order to be able to associate ids
2. shuffle queries
3. execute all queries in shuffled order (but store measurements under original ID)
4. goto 2

File: src/random_read_worker.rs (hunk 4 above)
Comment (liss-h) [reply src/random_read_worker.rs]: added comment

File: src/main.rs (hunk 2 above)
Comment (nkaralis) [reply src/main.rs]: I think in IGUANA, we refer to it as QPS, right @bigerl?

File: src/random_read_worker.rs (hunk 5 above)
Comment (nkaralis) [reply src/random_read_worker.rs:62]: Not something to address in this PR, but having a fixed order might also come in handy at some point.

File: src/main.rs (hunk 2 above)
Comment (bigerl) [reply src/main.rs]: regarding the hole (time -> inv -> avg) vs (time -> avg -> inv) discussion, Charlotte recommended this: https://math.stackexchange.com/questions/3107661/why-is-the-inverse-of-an-average-of-numbers-not-the-same-as-the-average-of-the-i 

tldr: it's not the same

I don't have the time right now to completely wrap my head around it, though. So, no real recommendation from my side. 

For how IGUANA does it, I summon @nck-mlcnv :)

File: src/random_read_worker.rs (hunk 5 above)
Comment (liss-h) [reply src/random_read_worker.rs:62]: Created an issue

File: src/main.rs (hunk 2 above)
Comment (liss-h) [reply src/main.rs]: I have now implemented the time -> avg -> inv because supposedly iguana does it that way. Could you double check, this is what we wanted/its correct, @nkaralis?

Comment (nkaralis) [reply src/main.rs]: looks good

**Comment by nkaralis** [comment]
Creating a separate file for the error handling makes sense.

I have some trouble following some parts due to my limited knowledge in rust, but, overall, it looks good.

**Comment by liss-h** [comment]
> Creating a separate file for the error handling makes sense.
> 
> I have some trouble following some parts due to my limited knowledge in rust, but, overall, it looks good.

Just to have it documented here: we had a quick chat and cleared up some things
